### ISSUE TITLE
[no-extra-semi]: Unnecessary semicolon when semicolon necessary

### ISSUE THREAD

**Issue by peterjuras** [eslint/eslint#4457]
Hi.

We're using eslint with babel and I found a case where the 'no-extra-semi' rule gets fired although the semicolon is necessary for the code to remain correct. The following code throws the warning:

````js
export default class SemiTest {
  foo() { return 'bar'; }
}; // ESLINT-WARNING: Unncessary semicolon

['one', 'two', 'three'].forEach(value => console.log(value));
````
However, if the semicolon is removed the code will no longer work correctly because the array declaration is treated like the array-access syntax and the code breaks. 

This probably due to whats happening in babel with the export default transformation and JavaScript trying to access a certain property of that object.

Edit:
ESLint version: 1.9.0
ESLint configuration:
````
{
  "parser": "babel-eslint",
  "rules": {
    "no-extra-semi": 1
  }
}
````
Full ESLint output:
````
/Users/juras/Desktop/semi/semi.js
  5:2  warning  Unnecessary semicolon  no-extra-semi

✖ 1 problem (0 errors, 1 warning)
````
Expectation: No warning about unecessary semicolon.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by mysticatea** [comment]
Thank you for reporting!

I think this is correct behavior because that class is a declaration (not an expression).
If you enclose the class declaration with parentheses, it becomes an expression. Then it needs a semicolon.

```js
export default (class SemiTest {
  foo() { return 'bar'; }
});
```

This is similar to:

```js
function foo() {
}
['one', 'two', 'three'].forEach(value => console.log(value)); // OK
```

```js
(function foo() {
})
['one', 'two', 'three'].forEach(value => console.log(value)); // Error!
```

**Comment by peterjuras** [comment]
Thanks, I added parantheses and it works as expected now.

**Comment by michaelficarra** [comment]
Woah, you don't want to add the parentheses, you want to remove the semicolon. It's unnecessary.

**Comment by nzakas** [comment]
A declaration doesn't need a semicolon, it should work fine by removing it. If not, then please file a bug with Babel.
